wrangler-service/src/main/java/co/cask/wrangler/service/spanner/SpannerService.java

```diff
@@ -0,0 +1,400 @@
+      responder.sendJson(new SpannerResponse<>(databases));
+    } catch (BadRequestException e) {
+      responder.sendError(HttpURLConnection.HTTP_BAD_REQUEST, e.getMessage());
+    } catch (Exception e) {
+      responder.sendError(HttpURLConnection.HTTP_INTERNAL_ERROR, e.getMessage());
+    }
+  }
+
+  /**
+   * Lists spanner tables for a spanner database
+   *
+   * @param request HTTP Request handler.
+   * @param responder HTTP Response handler.
+   */
+  @GET
+  @Path("/connections/{connection-id}/spanner/instances/{instance-id}/databases/{database-id}/tables")
+  public void getSpannerTables(HttpServiceRequest request, HttpServiceResponder responder,
+                               @PathParam("connection-id") final String connectionId,
+                               @PathParam("instance-id") final String instanceId,
+                               @PathParam("database-id") final String databaseId) {
+    try {
+      Connection connection = store.get(connectionId);
+      validateConnection(connectionId, connection);
+      List<SpannerTable> tables = getTables(connection, instanceId, databaseId);
+      responder.sendJson(new SpannerResponse<>(tables));
+    } catch (BadRequestException e) {
+      responder.sendError(HttpURLConnection.HTTP_BAD_REQUEST, e.getMessage());
+    } catch (Exception e) {
+      responder.sendError(HttpURLConnection.HTTP_INTERNAL_ERROR, e.getMessage());
+    }
+  }
+
+  /**
+   * Read spanner table into a workspace and return the workspace identifier
+   */
+  @GET
+  @Path("/connections/{connection-id}/spanner/instances/{instance-id}/databases/{database-id}/tables/{table-id}/read")
+  public void readTable(HttpServiceRequest request, HttpServiceResponder responder,
+                        @PathParam("connection-id") final String connectionId,
+                        @PathParam("instance-id") final String instanceId,
+                        @PathParam("database-id") final String databaseId,
+                        @PathParam("table-id") final String tableId,
+                        @QueryParam("scope") @DefaultValue(WorkspaceDataset.DEFAULT_SCOPE) String scope,
+                        @QueryParam("limit") @DefaultValue(DEFAULT_ROW_LIMIT) String limit) {
+    try {
+      Connection connection = store.get(connectionId);
+      validateConnection(connectionId, connection);
+      Schema schema = getTableSchema(connection, instanceId, databaseId, tableId);
+      List<Row> data = getTableData(connection, instanceId, databaseId, tableId, schema, Long.parseLong(limit));
+
+      // create workspace id
+      String identifier = ServiceUtils.generateMD5(String.format("%s:%s", scope, tableId));
+      ws.createWorkspaceMeta(identifier, scope, tableId);
+
+      // write data to workspace
+      ObjectSerDe<List<Row>> serDe = new ObjectSerDe<>();
+      byte[] dataBytes = serDe.toByteArray(data);
+      ws.writeToWorkspace(identifier, WorkspaceDataset.DATA_COL, DataType.RECORDS, dataBytes);
```

Comment:
this means all the records are stored into a single cell doesn't it... I think we'll need to create a new backwards incompatible version of dataprep at some point.

Reply:
yes, also the data lifecycle is not clearly defined.